sense platform/sensors/CSNoiseSensor.m

```diff
@@ -369,23 +366,32 @@ - (void) dealloc {
     [NSThread sleepForTimeInterval: 0.1];
 }
 
+/** 
+ When receiving new screen state data, we do the following:
+  - When the screen gets locked we allow new recordings by setting screenstateBlocksRecording to NO
+  - When the screen gets unlocked we stop any running recordings and blocks future recordings by setting screenstateBlocksRecording to YES
+  - When the screen gets an on/off switch event (and we don't know if it is on or off) we take a conservative approach and stop any recording, and cancel the next recording as well (because there is some delay in lock and unlock detections).
+ */
 - (void) onNewData:(NSNotification*)notification {
-    NSString* sensor = notification.object;
-    if ([sensor isEqualToString:kCSSENSOR_SCREEN_STATE]) {
-        // if receive dispaly event and the sampleonly flag is no do nothing otherwise call schedule but
-        // if you record stop
-        NSString* json = [notification.userInfo valueForKey:@"value"];
-        NSString *screenState = [json valueForKey:@"screen"];
-        
-        // keep track of the state of the screen
-        if ([screenState isEqualToString:@"off"])
-            screenIsOn = NO;
-        else {
-            screenIsOn = YES;
+    if ([notification.object isEqualToString:kCSSENSOR_SCREEN_STATE]) {
+           NSString *screenState = [[notification.userInfo valueForKey:@"value"] valueForKey:@"screen"];
+
+		if ([screenState isEqualToString:kVALUE_IDENTIFIER_SCREEN_LOCKED]) {
+            screenstateBlocksRecording = NO;
+			nextRecordingCancelled = NO;
+		} else if ([screenState isEqualToString:kVALUE_IDENTIFIER_SCREEN_UNLOCKED]){
+            screenstateBlocksRecording = YES;
             if ([audioRecorder isRecording] == YES) {
                 [audioRecorder stop];
             }
-        }
+		} else if ([screenState isEqualToString:kVALUE_IDENTIFIER_SCREEN_ONOFF_SWITCH]) {
+			if ([audioRecorder isRecording] == YES) {
```

Comment:
@jorishjanssen details details, but I think this will not be thread-safe:
- startRecording checks screenStateBlocksRecording == NO
- onNewData is called, sets screenStateBlocksRecording = YES, checks adioRecorder isRecording == NO, ends
- startRecoding gets control back, calls startRecording
- user gets a red bar :(